node/src/components/consensus/protocols/highway/performance_meter.rs

```diff
@@ -0,0 +1,137 @@
+            current_round_len: round_len,
+            last_switch_round_id: current_round_id,
+        }
+    }
+
+    pub fn current_round_len(&self) -> TimeDiff {
+        self.current_round_len
+    }
+
+    /// Whenever a block is proposed in a round and nodes cast their confirmation and witness
+    /// units, a "max quorum" can be calculated for each node - the maximum quorum for which there
+    /// exists a level-1 summit within the round, containing the particular node.
+    /// This function calculates the average max quorum for this node out of `BLOCKS_TO_CONSIDER`
+    /// most recent ancestor proposals of the current fork choice.
+    /// If this average max quorum is below a certain threshold, a higher round length will be
+    /// returned.
+    /// If it is above a certain threshold, and the current round ID is divisible by a certain
+    /// number, a lower round length is returned.
+    pub fn calculate_new_length<C: Context>(&mut self, state: &State<C>) -> TimeDiff {
+        let panorama = state.panorama();
+        let latest_block = match state.fork_choice(panorama) {
+            Some(block) => block,
+            None => {
+                // we have no blocks to check - just return the current setting
+                return self.current_round_len;
+            }
+        };
+
+        let blocks_to_check = state.ancestor_hashes(latest_block);
+
+        let max_quora: Vec<_> = blocks_to_check
+            .take_while(|block| state.unit(block).round_id() >= self.last_switch_round_id)
+            .filter_map(|block| {
+                let round_id = state.unit(block).round_id();
+                (!matches!(
+                    round_participation(state, &panorama[self.our_vid], round_id),
+                    RoundParticipation::Unassigned
+                ))
+                .then(|| {
+                    let (assigned_weight, latest_units) =
+                        assigned_weight_and_latest_unit(state, panorama, round_id);
+                    let max_quorum = find_max_quora(state, block, &latest_units)
+                        .get(self.our_vid)
+                        .copied()
+                        .unwrap_or(0u64.into());
+                    max_quorum.0 as f64 / assigned_weight.0 as f64
+                })
+            })
+            .take(BLOCKS_TO_CONSIDER)
+            .collect();
+
+        if max_quora.len() < BLOCKS_TO_CONSIDER {
+            return self.current_round_len;
+        }
+
+        let avg_max_quorum = max_quora.iter().sum::<f64>() / max_quora.len() as f64;
+
+        let current_round_id = state.unit(latest_block).round_id();
+        let current_round_index = round_index(current_round_id, self.current_round_len);
+
+        #[allow(clippy::integer_arithmetic)]
+        if avg_max_quorum < SLOW_DOWN_THRESHOLD {
+            self.current_round_len = min(self.current_round_len * 2, self.max_round_len);
+            self.last_switch_round_id = current_round_id;
```

Comment:
Should this line still be called even if `self.current_round_len` didn't change in the line above (i.e. if we're already running at `max_round_len`)?  Same question for the similar call on line 122.  I guess it probably should, so we don't keep considering ever-increasing numbers of block when calculating `max_quora`, but if we ever expose this field (e.g. to the status endpoint) it's a bit misleading maybe?

Reply:
Good point. I think we might omit this line, actually. The number of blocks considered won't be ever increasing, anyway, because we take at most `BLOCKS_TO_CONSIDER` of them, and it is set to just 5 currently (which is probably quite enough).

Reply:
> The number of blocks considered won't be ever increasing, anyway

Ah - of course! :)